share/shwap/range_namespace_data.go

```diff
@@ -0,0 +1,323 @@
+	// - The range ends before the row ends
+	// - AND spans multiple rows
+	endProof := endsMidRow && isMultiRow
+
+	rngData := RangeNamespaceData{
+		Start:  from.Row,
+		Shares: make([][]libshare.Share, numRows),
+		Proof:  make([]*Proof, numRows),
+	}
+	for i, row, col := 0, from.Row, from.Col; i < numRows; i++ {
+		rowShares := shares[i]
+		// end index will be explicitly set only for the last row in range.
+		// in other cases, it will be equal to the odsSize.
+		exclusiveEnd := odsSize
+		if i == len(shares)-1 {
+			// `to.Col` is an inclusive index
+			exclusiveEnd = to.Col + 1
+		}
+
+		// ensure that all shares from the range belong to the requested namespace
+		for offset := range rowShares[col:exclusiveEnd] {
+			if !namespace.Equals(rowShares[from.Col+offset].Namespace()) {
+				return RangeNamespaceData{},
+					fmt.Errorf("targeted namespace was not found in share at {Row: %d, Col: %d}",
+						row, col+offset,
+					)
+			}
+		}
+
+		rngData.Shares[i] = rowShares[col:exclusiveEnd]
+		rngData.Proof[i] = &Proof{
+			rowRootProof: rowRootProofs[row],
+			shareProof:   &nmt.Proof{},
+		}
+
+		// reset from.Col as we are moving to the next row.
+		col = 0
+		row++
+	}
+	// incomplete from.Col needs a proof for the first row to be computed
+	if startProof {
+		endCol := odsSize
+		if from.Row == to.Row {
+			endCol = to.Col + 1
+		}
+		sharesProofs, err := generateSharesProofs(from.Row, from.Col, endCol, odsSize, namespace, shares[0])
+		if err != nil {
+			return RangeNamespaceData{}, fmt.Errorf("failed to generate proof for row %d: %w", from.Row, err)
+		}
+		rngData.Proof[0] = &Proof{shareProof: sharesProofs, rowRootProof: rowRootProofs[from.Row]}
+	}
+	// incomplete to.Col needs a proof for the last row to be computed
+	if endProof {
+		sharesProofs, err := generateSharesProofs(to.Row, 0, to.Col+1, odsSize, namespace, shares[numRows-1])
+		if err != nil {
+			return RangeNamespaceData{}, fmt.Errorf("failed to generate proof for row %d: %w", to.Row, err)
+		}
+		rngData.Proof[numRows-1] = &Proof{shareProof: sharesProofs, rowRootProof: rowRootProofs[to.Row]}
+	}
+	return rngData, nil
+}
+
+// Verify verifies the underlying shares are included in the data root.
+func (rngdata *RangeNamespaceData) Verify(
+	namespace libshare.Namespace,
+	from SampleCoords,
+	to SampleCoords,
+	dataHash []byte,
+) error {
+	return rngdata.VerifyShares(rngdata.Shares, namespace, from, to, dataHash)
+}
+
+// VerifyShares verifies the passed shares are included in the data root.
+// `[][]libshare.Share` is a collection of the row shares from the range
+// NOTE: the underlying shares will be ignored even if they are not empty.
+func (rngdata *RangeNamespaceData) VerifyShares(
+	shares [][]libshare.Share,
+	namespace libshare.Namespace,
+	from SampleCoords,
+	to SampleCoords,
+	dataHash []byte,
+) error {
+	proofs := rngdata.Proof
+	// compute the proofs for the complete rows
+	for i, row := 0, from.Row; i < len(shares); i++ {
+		// skip the incomplete rows
```

Comment:
you can additionally check that row root proof is not empty here

Reply:
whats the purpose for adding the additional check?

Reply:
you are doing additional computation here that takes some time and memory. but they won't make any sense if rowRootProofs were not set correctly as it will fail on this step. Actually, I meant before the loop

Reply:
actually, we have to check it inside the loop for every proof[i], the corresponding row root proof is set.